### PR TITLE
HIVE-25621: Added HivePrivilegeObjects for the commands Alter table m…

### PR DESCRIPTION
**Pull request by saihemanth-cloudera**
…erge/concatenate

### What changes were proposed in this pull request?
Added HivePrivilegeObjects for alter table merge/concatenate commands.

### Why are the changes needed?
To prevent  malicious users from doing alter operations on table.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Local machine, remote cluster.

### CODE REVIEW

**Comment by saihemanth-cloudera** [comment]
@dengzhhu653 - Can you please review my patch?

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/storage/compact/AlterTableCompactAnalyzer.java [hunk 1]
```diff
@@ -67,6 +73,17 @@ protected void analyzeCommand(TableName tableName, Map<String, String> partition
...
     AlterTableCompactDesc desc = new AlterTableCompactDesc(tableName, partitionSpec, type, isBlocking, mapProp);
+    Table table = getTable(tableName);
+    WriteEntity.WriteType writeType = null;
+    if (AcidUtils.isTransactionalTable(table)) {
+      setAcidDdlDesc(desc);
+      writeType = WriteType.DDL_EXCLUSIVE;
+    } else {
+      writeType = WriteEntity.determineAlterTableWriteType(AlterTableType.COMPACT);
+    }
+    inputs.add(new ReadEntity(table));
```
Comment (dengzhhu653) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/storage/compact/AlterTableCompactAnalyzer.java:84]: should we take care of `partitionSpec` as well?

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/storage/compact/AlterTableCompactAnalyzer.java [hunk 2]
```diff
@@ -67,6 +73,17 @@ protected void analyzeCommand(TableName tableName, Map<String, String> partition
     }
 
     AlterTableCompactDesc desc = new AlterTableCompactDesc(tableName, partitionSpec, type, isBlocking, mapProp);
+    Table table = getTable(tableName);
+    WriteEntity.WriteType writeType = null;
+    if (AcidUtils.isTransactionalTable(table)) {
+      setAcidDdlDesc(desc);
+      writeType = WriteType.DDL_EXCLUSIVE;
```
Comment (dengzhhu653) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/storage/compact/AlterTableCompactAnalyzer.java:80]: cloud you please explain a little bit why we choose DDL_EXCLUSIVE for transactional table?  does it works the same for insert only tables?

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/storage/concatenate/AlterTableConcatenateAnalyzer.java [hunk 3]
```diff
@@ -95,10 +97,14 @@ protected void analyzeCommand(TableName tableName, Map<String, String> partition
     }
   }
 
-  private void compactAcidTable(TableName tableName, Map<String, String> partitionSpec) throws SemanticException {
+  private void compactAcidTable(TableName tableName, Table table, Map<String, String> partitionSpec) throws SemanticException {
     boolean isBlocking = !HiveConf.getBoolVar(conf, ConfVars.TRANSACTIONAL_CONCATENATE_NOBLOCK, false);
 
     AlterTableCompactDesc desc = new AlterTableCompactDesc(tableName, partitionSpec, "MAJOR", isBlocking, null);
+    WriteEntity.WriteType writeType = WriteEntity.WriteType.DDL_EXCLUSIVE;
```
Comment (dengzhhu653) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/storage/concatenate/AlterTableConcatenateAnalyzer.java:104]: should we take care of `partitionSpec` as well?

**Comment by saihemanth-cloudera** [comment]
@dengzhhu653 - Can you please review my patch again? Thanks

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/storage/compact/AlterTableCompactAnalyzer.java [hunk 4]
```diff
@@ -67,6 +73,18 @@ protected void analyzeCommand(TableName tableName, Map<String, String> partition
     }
 
     AlterTableCompactDesc desc = new AlterTableCompactDesc(tableName, partitionSpec, type, isBlocking, mapProp);
+//    Table table = getTable(tableName);
```
Comment (dengzhhu653) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/storage/compact/AlterTableCompactAnalyzer.java:76]: nit: can we remove these comments as well?

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/storage/compact/AlterTableCompactAnalyzer.java [hunk 5]
```diff
@@ -27,7 +27,13 @@
 import org.apache.hadoop.hive.ql.ddl.DDLWork;
 import org.apache.hadoop.hive.ql.ddl.DDLSemanticAnalyzerFactory.DDLType;
 import org.apache.hadoop.hive.ql.ddl.table.AbstractAlterTableAnalyzer;
+import org.apache.hadoop.hive.ql.ddl.table.AlterTableType;
 import org.apache.hadoop.hive.ql.exec.TaskFactory;
+import org.apache.hadoop.hive.ql.hooks.ReadEntity;
```
Comment (dengzhhu653) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/storage/compact/AlterTableCompactAnalyzer.java:32]: nit: worthless imports

File: ql/src/test/results/clientpositive/llap/acid_insert_overwrite_update.q.out [hunk 6]
```diff
@@ -243,16 +247,28 @@ POSTHOOK: Input: default@orc_test_txn@year=2018
...
 PREHOOK: type: ALTERTABLE_COMPACT
+PREHOOK: Input: default@orc_test_txn
+PREHOOK: Output: default@orc_test_txn
 POSTHOOK: query: alter table orc_test_txn partition(year='2017') compact 'major'
 POSTHOOK: type: ALTERTABLE_COMPACT
+POSTHOOK: Input: default@orc_test_txn
+POSTHOOK: Output: default@orc_test_txn
 PREHOOK: query: alter table orc_test_txn partition(year='2018') compact 'major'
 PREHOOK: type: ALTERTABLE_COMPACT
+PREHOOK: Input: default@orc_test_txn
```
Comment (dengzhhu653) [inline ql/src/test/results/clientpositive/llap/acid_insert_overwrite_update.q.out:266]: why does partition 'year=2018' not present in the Input/Output?

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/storage/compact/AlterTableCompactAnalyzer.java (hunk 4 above)
Comment (saihemanth-cloudera) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/storage/compact/AlterTableCompactAnalyzer.java:76]: Sure

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/storage/compact/AlterTableCompactAnalyzer.java (hunk 5 above)
Comment (saihemanth-cloudera) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/storage/compact/AlterTableCompactAnalyzer.java:32]: Ack

File: ql/src/test/results/clientpositive/llap/acid_insert_overwrite_update.q.out (hunk 6 above)
Comment (saihemanth-cloudera) [reply ql/src/test/results/clientpositive/llap/acid_insert_overwrite_update.q.out:266]: Ack

**Comment by saihemanth-cloudera** [comment]
@dengzhhu653 Can you review my patch again?

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/storage/concatenate/AlterTableConcatenateDesc.java [hunk 7]
```diff
@@ -44,7 +47,8 @@ public class AlterTableConcatenateDesc implements DDLDesc {
   private final TableDesc tableDesc;
 
   public AlterTableConcatenateDesc(TableName tableName, Map<String, String> partitionSpec, ListBucketingCtx lbCtx,
-      Path inputDir, Path outputDir, Class<? extends InputFormat> inputFormatClass, TableDesc tableDesc) {
+      Path inputDir, Path outputDir, Class<? extends InputFormat> inputFormatClass, TableDesc tableDesc) throws SemanticException {
+    super(AlterTableType.COMPACT, tableName, partitionSpec, null, false, false, null);
```
Comment (dengzhhu653) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/storage/concatenate/AlterTableConcatenateDesc.java:51]: should this be AlterTableType.MERGEFILES other than COMPACT?

**Comment by dengzhhu653** [comment]
> @dengzhhu653 Can you review my patch again?

leave a minor comments, also the failed tests seem related, others look good to me.

File: ql/src/test/results/clientpositive/llap/orc_merge5.q.out [hunk 8]
```diff
@@ -385,6 +385,7 @@ STAGE PLANS:
   Stage: Stage-0
     Concatenate
       table name: default.orc_merge5b_n0
+      table name: default.orc_merge5b_n0
```
Comment (dengzhhu653) [inline ql/src/test/results/clientpositive/llap/orc_merge5.q.out:388]: nit: there seems duplicate, the same outputs twice

**Review by dengzhhu653 — APPROVED** [review]

File: ql/src/test/results/clientpositive/llap/orc_merge5.q.out (hunk 8 above)
Comment (saihemanth-cloudera) [reply ql/src/test/results/clientpositive/llap/orc_merge5.q.out:388]: No, This is not a duplicate. Input and output might be outputted here. If it was a duplicate, we would have had an error in the output file.

Comment (dengzhhu653) [reply ql/src/test/results/clientpositive/llap/orc_merge5.q.out:388]: I wondering the two one is from AlterTableConcatenateDesc,
https://github.com/apache/hive/blob/7b3ecf617a6d46f48a3b6f77e0339fd4ad95a420/ql/src/java/org/apache/hadoop/hive/ql/ddl/table/storage/concatenate/AlterTableConcatenateDesc.java#L57-L60
the other comes from AbstractAlterTableDesc:
https://github.com/apache/hive/blob/7b3ecf617a6d46f48a3b6f77e0339fd4ad95a420/ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableDesc.java#L64-L67

Comment (saihemanth-cloudera) [reply ql/src/test/results/clientpositive/llap/orc_merge5.q.out]: Ack

**Comment by saihemanth-cloudera** [comment]
@dengzhhu653 - Tests are green now. Can you do another review here?

**Review by dengzhhu653 — APPROVED** [review]

**Pull request merged**